compiler/src/main/kotlin/io/github/aplcornell/viaduct/circuitcodegeneration/BackendCodeGenerator.kt

```diff
@@ -52,18 +53,17 @@ private class BackendCodeGenerator(
             PropertySpec.builder("runtime", ViaductRuntime::class).initializer("runtime")
                 .addModifiers(KModifier.PRIVATE).build(),
         )
-        // TODO Replace with a call to ProtocolAnalysis
-        val participatingProtocols = program.circuits.map { it.protocol.value }
-        for (protocol in participatingProtocols) {
+        for (protocol in program.protocols()) {
             for (property in codeGenerator.setup(protocol)) {
                 classBuilder.addProperty(property)
             }
         }
-        for (circuit in program.circuits) {
-            classBuilder.addFunction(generate(circuit)).build()
-        }
-        for (function in program.functions) {
-            classBuilder.addFunction(generate(function)).build()
+        for (declaration in program.declarations) {
```

Comment:
You should skip circuits if the current host is uninvolved with the protocol. Not sure about functions...

Reply:
Yep, working on this and the above now